### PR TITLE
Async

### PR DESCRIPTION
**Pull request by matemax**

### CODE REVIEW

**Comment by matemax** [comment]
async quality

File: examples/sdk_examples/image_rotation.py [hunk 1]
```diff
@@ -31,5 +32,19 @@ def rotateNEstimateImage():
...
+    """
+    Async example of image rotation.
+    """
+    nonRotatedImage = VLImage.load(filename=EXAMPLE_O)
+    faceEngine = VLFaceEngine()
+    orientationModeEstimator = faceEngine.createOrientationModeEstimator()
+    #: rotate & estimate | not rotated
+    image = VLImage.rotate(nonRotatedImage, RotationAngle.ANGLE_0)
+    orientation = await orientationModeEstimator.estimate(image, asyncEstimate=True)
+    pprint.pprint(orientation)
```
Comment (matemax) [inline examples/sdk_examples/image_rotation.py:45]: + get example

File: lunavl/sdk/detectors/facedetector.py [hunk 2]
```diff
@@ -75,6 +78,14 @@ def __init__(self, coreLandmark68: CoreLandmarks68):
         super().__init__(coreLandmark68)
 
 
+def postProcessing(error, detectRes, image):
```
Comment (matemax) [inline lunavl/sdk/detectors/facedetector.py]: .

File: lunavl/sdk/detectors/facedetector.py [hunk 3]
```diff
@@ -124,6 +135,57 @@ def asDict(self) -> Dict[str, Any]:
...
+        res.append(
+            [
+                FaceDetection(coreDetection, vlImage) if coreDetection.isValid() else None
+                for coreDetection in imagesDetections
+            ]
+        )
+    return res
+
+
+def postProcessingRedetect(error, detectionsBatch, coreImages, images):
```
Comment (matemax) [inline lunavl/sdk/detectors/facedetector.py]: .

File: lunavl/sdk/detectors/humandetector.py [hunk 4]
```diff
@@ -93,6 +96,82 @@ def asDict(self) -> Dict[str, Any]:
         return res
 
 
+def createHumanDetection(image: VLImage, detection: Detection, landmarks17: Optional[CoreLandmarks17]):
```
Comment (matemax) [inline lunavl/sdk/detectors/humandetector.py:100]: .

File: lunavl/sdk/detectors/humandetector.py [hunk 5]
```diff
@@ -93,6 +96,82 @@ def asDict(self) -> Dict[str, Any]:
...
+    if landmarks17:
+        human.landmarks17_opt.set(landmarks17)
+    return HumanDetection(human, image)
+
+
+def collectDetectionsResult(
+    fsdkDetectRes,
+    images: Union[List[Union[VLImage, ImageForDetection]], List[ImageForRedetection]],
+    detectLandmarks: bool = True,
+):
```
Comment (matemax) [inline lunavl/sdk/detectors/humandetector.py]: .

File: lunavl/sdk/errors/errors.py [hunk 6]
```diff
@@ -55,85 +56,99 @@ def __repr__(self) -> str:
...
+    BatchedInternalError = ErrorInfo(100035, "Batching error", "{}")
+
+    CreationDescriptorError = ErrorInfo(110001, "Creation descriptor error", "{}")
+    CreationBatchDescriptorsError = ErrorInfo(110002, "Creation descriptor error", "{}")
+    CreationImageError = ErrorInfo(110003, "Creation core image error", "{}")
+    EstimationDescriptorError = ErrorInfo(110004, "Estimation descriptor error", "{}")
+    EstimationBatchDescriptorError = ErrorInfo(110005, "Estimation descriptor error", "{}")
+    EstimationBasicAttributeError = ErrorInfo(110006, "Estimation basic attributes error", "{}")
+    BatchEstimationBasicAttributeError = ErrorInfo(110007, "Batch estimation basic attributes error", "{}")
+    EstimationAGSError = ErrorInfo(110008, "Estimation AGS error", "{}")
```
Comment (matemax) [inline lunavl/sdk/errors/errors.py]: .

File: lunavl/sdk/estimators/estimators_utils/extractor_utils.py [hunk 7]
```diff
@@ -1,21 +1,43 @@
...
 
+from lunavl.sdk.async_task import AsyncTask
 from lunavl.sdk.descriptors.descriptors import BaseDescriptor, BaseDescriptorFactory, BaseDescriptorBatch
 from lunavl.sdk.errors.errors import LunaVLError
 from lunavl.sdk.errors.exceptions import LunaSDKException, assertError
 from ..body_estimators.humanwarper import HumanWarp, HumanWarpedImage
 from ..face_estimators.facewarper import FaceWarp, FaceWarpedImage
 
 
+def postProcessing(error, gs, descriptor):
```
Comment (matemax) [inline lunavl/sdk/estimators/estimators_utils/extractor_utils.py]: .

File: lunavl/sdk/estimators/face_estimators/credibility.py [hunk 8]
```diff
@@ -84,6 +84,11 @@ def asDict(self) -> dict:
         return {"estimations": {"score": self.score}, "prediction": self.prediction.value}
 
 
+def postProcessing(error, credibility):
```
Comment (matemax) [inline lunavl/sdk/estimators/face_estimators/credibility.py]: .

File: lunavl/sdk/estimators/face_estimators/livenessv1.py [hunk 9]
```diff
@@ -105,6 +105,17 @@ def quality(self) -> float:
         return self._coreEstimation.qualityScore
 
 
+def postProcessing(error, estimation):
```
Comment (matemax) [inline lunavl/sdk/estimators/face_estimators/livenessv1.py]: .

File: lunavl/sdk/async_task.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+from lunavl.sdk.errors.exceptions import assertError
+
+TaskResult = TypeVar("TaskResult")
+CoreTaskResult = TypeVar("CoreTaskResult")
+
+
+def postProcessingBatch(
+    error: FSDKErrorResult, coreEstimations: List[CoreTaskResult], resultClass: Type[TaskResult]
+) -> List[TaskResult]:
```
Comment (matemax) [inline lunavl/sdk/async_task.py:13]: .

**Comment by matemax** [comment]
need tests

File: pyproject.toml [hunk 11]
```diff
@@ -17,9 +17,9 @@ Pillow = {version = "*", markers = "sys_platform == 'win32'"}
 pillow-simd = {version = "*", markers = "sys_platform != 'win32'"}
 dataclasses = {version="*", markers="python_version == '3.6'"}
 
-#[tool.poetry.dependencies.FaceEngine]
-#git = "ssh://git@git.visionlabs.ru/face-recognition/FaceEnginePythonBindings.git"
-#tag = "v.5.5.1.0"
+# [tool.poetry.dependencies.FaceEngine]
+# git = "ssh://git@git.visionlabs.ru/face-recognition/FaceEnginePythonBinding.git"
+# tag = "5.6.0"
```
Comment (gluhar2006) [inline pyproject.toml]:
```suggestion
# tag = "5.6.0.0"
```

File: pyproject.toml [hunk 12]
```diff
@@ -17,9 +17,9 @@ Pillow = {version = "*", markers = "sys_platform == 'win32'"}
 pillow-simd = {version = "*", markers = "sys_platform != 'win32'"}
 dataclasses = {version="*", markers="python_version == '3.6'"}
 
-#[tool.poetry.dependencies.FaceEngine]
-#git = "ssh://git@git.visionlabs.ru/face-recognition/FaceEnginePythonBindings.git"
-#tag = "v.5.5.1.0"
+# [tool.poetry.dependencies.FaceEngine]
+# git = "ssh://git@git.visionlabs.ru/face-recognition/FaceEnginePythonBinding.git"
```
Comment (gluhar2006) [inline pyproject.toml]:
```suggestion
# git = "ssh://git@git.visionlabs.ru/face-recognition/FaceEnginePythonBindings.git"
```

File: lunavl/sdk/async_task.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    return [resultClass(coreEstimation) for coreEstimation in coreEstimations]
+
+
+def postProcessingBatchWithAggregation(
+    error: FSDKErrorResult,
+    coreEstimations: List[CoreTaskResult],
+    aggregatedAttribute: CoreTaskResult,
+    resultClass: Type[TaskResult],
+    aggregate: bool,
+) -> tuple[List[TaskResult], Optional[TaskResult]]:
```
Comment (gluhar2006) [inline lunavl/sdk/async_task.py]: `tuple`?`
pyproject.toml: python = ">=3.6"

File: lunavl/sdk/async_task.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        aggregatedAttribute: aggregated core estimation
+        aggregate: need or not aggregate result
+        resultClass: result class
+
+    Returns:
+        list of lunavl structure based on core estimation + aggregated estimation
+    """
+    assertError(error)
+
+    estimations = [resultClass(coreEstimation) for coreEstimation in coreEstimations]
```
Comment (gluhar2006) [inline lunavl/sdk/async_task.py:50]: ```estimations = postProcessingBatch(...)```?

File: lunavl/sdk/async_task.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    if aggregate:
+        return estimations, resultClass(aggregatedAttribute)
+    else:
+        return estimations, None
+
+
+def postProcessing(error: FSDKErrorResult, coreEstimation: CoreTaskResult, resultClass: Type[TaskResult]) -> TaskResult:
```
Comment (gluhar2006) [inline lunavl/sdk/async_task.py:57]: comments

File: lunavl/sdk/async_task.py [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+    """
+    Default estimation post processing factory.
+
+    Factory is intended for generation standard post procesing  core estimation object (conversion to lunavl structure
+    and check errors)
+
+    Attributes:
+        resultClass: result class
+    """
+
```
Comment (gluhar2006) [inline lunavl/sdk/async_task.py:81]:
```suggestion
    __slots__ = ("resultClass",)

```

File: lunavl/sdk/detectors/base.py [hunk 17]
```diff
@@ -173,3 +173,40 @@ def validateBatchDetectInput(
     else:
         errors.append(LunaVLError.Ok.format(LunaVLError.Ok.description))
     raise LunaSDKException(LunaVLError.BatchedInternalError.format(LunaVLError.fromSDKError(mainError).detail), errors)
+
+
+def validateReDetectInput(detector, coreImages: List[CoreImage], detectAreas: List[List[Detection]]):
```
Comment (gluhar2006) [inline lunavl/sdk/detectors/base.py:178]: wrong typing: List[CoreImage] -> Union[CoreImage, List[CoreImage]], etc

File: lunavl/sdk/detectors/facedetector.py [hunk 18]
```diff
@@ -124,6 +128,165 @@ def asDict(self) -> Dict[str, Any]:
...
+) -> List[List[Optional[FaceDetection]]]:
+    """
+    Collect redetection results from core reply and prepare face detections
+    Args:
+        fsdkDetectRes: fsdk (re)detect results
+        images: incoming images
+    Returns:
+        return list of lists detection, order of detection lists is corresponding to order input images
+    """
+    return _collectDetectionsResult(fsdkDetectRes=fsdkDetectRes, images=images, isRedectResult=True)
```
Comment (gluhar2006) [inline lunavl/sdk/detectors/facedetector.py:196]: partial? here and below

File: lunavl/sdk/detectors/facedetector.py [hunk 19]
```diff
@@ -124,6 +128,165 @@ def asDict(self) -> Dict[str, Any]:
...
+def postProcessingOne(error: FSDKErrorResult, detectRes: Face, image: VLImage) -> Optional[FaceDetection]:
+    """
+    Convert core face detection to `FaceDetection` after detect one and error check.
+
+    Args:
+        error: detection error, usually error.isError is False
+        detectRes: detections
+        image: original image
+
+    Raises:
```
Comment (gluhar2006) [inline lunavl/sdk/detectors/facedetector.py:223]: raises? and below

File: lunavl/sdk/detectors/humandetector.py [hunk 20]
```diff
@@ -93,6 +97,165 @@ def asDict(self) -> Dict[str, Any]:
         return res
 
 
+def createHumanDetection(image: VLImage, detection: Detection, landmarks17: Optional[CoreLandmarks17]):
```
Comment (gluhar2006) [inline lunavl/sdk/detectors/humandetector.py:100]:
```suggestion
def createHumanDetection(image: VLImage, detection: Detection, landmarks17: Optional[CoreLandmarks17] = None) -> HumanDetection:
```

File: lunavl/sdk/detectors/humandetector.py [hunk 21]
```diff
@@ -93,6 +97,165 @@ def asDict(self) -> Dict[str, Any]:
...
+
+    Raises:
+        LunaSDKException: if detect is failed
+        RuntimeError: if any detection is not valid
+    Returns:
+        human detection
+    """
+    assertError(error)
+    detectionResCount = detectRes.getSize()
+    if detectionResCount > 1 or detectionResCount == 0:
```
Comment (gluhar2006) [inline lunavl/sdk/detectors/humandetector.py:175]: not equal 1? seems that negative `detectionResCount` is not good result

File: lunavl/sdk/estimators/image_estimators/orientation_mode.py [hunk 22]
```diff
@@ -46,6 +46,18 @@ def __repr__(self):
         return self.value
 
 
+def postProcessingBatch(error, oientations):
```
Comment (gluhar2006) [inline lunavl/sdk/estimators/image_estimators/orientation_mode.py]: comments, typing

File: lunavl/sdk/luna_faces.py [hunk 23]
```diff
@@ -431,33 +433,65 @@ def detectOne(self, image: VLImage, detectArea: Optional[Rect] = None) -> Union[
             detectRes.coreEstimation, detectRes.image, self.estimatorsCollection, self._estimationSettings
         )
 
-    def detect(self, images: List[Union[VLImage, ImageForDetection]], limit: int = 5) -> List[List[VLFaceDetection]]:
+    def postProcessingDetectionBatch(self, detectRes: List[List[FaceDetection]]):
```
Comment (gluhar2006) [inline lunavl/sdk/luna_faces.py]:
```suggestion
    def postProcessingDetectionBatch(self, detectRes: List[List[Optional[FaceDetection]]]):
```

File: lunavl/sdk/luna_faces.py [hunk 24]
```diff
@@ -431,33 +433,65 @@ def detectOne(self, image: VLImage, detectArea: Optional[Rect] = None) -> Union[
...
+            limit: max number of detections per input image
+        Returns:
+            return list of lists detection, order of detection lists is corresponding to order of input images
+        """
+        detectRes = self._faceDetector.detect(images, limit, True, True)
+        return self.postProcessingDetectionBatch(detectRes)
+
+    def redetectOne(
+        self, image: Union[VLImage, VLFaceDetection], bBox: Rect
+    ) -> Union[Union[VLFaceDetection, None], Union[VLFaceDetection, None]]:
```
Comment (gluhar2006) [inline lunavl/sdk/luna_faces.py]: `Union[VLFaceDetection, None]` twice?

File: tests/base.py [hunk 25]
```diff
@@ -121,3 +122,43 @@ def getColorToImageMap() -> Dict[ColorFormat, VLImage]:
...
+        """
+        Assert batch async estimation with aggregation
+        Args:
+            task: async task
+            expectedTypeResult: expected type of result
+        """
+        isinstance(task, AsyncTask)
+        res = task.get()
+        isinstance(res, tuple)
+        assert 2 == len(res)
```
Comment (gluhar2006) [inline tests/base.py:166]: does not match the function name. Looks like there should be cycle

File: tests/test_facedetector.py [hunk 26]
```diff
@@ -468,3 +474,12 @@ def test_detect_one_by_area_outside_image(self):
...
                 with self.assertRaises(expected_exception=LunaSDKException):
                     detector.detectOne(image=VLIMAGE_ONE_FACE, detectArea=AREA_OUTSIDE_IMAGE)
+
+    def test_async_detect_face(self):
+        """
+        Test async detect face
+        """
+        task = self.defaultDetector.detectOne(VLIMAGE_ONE_FACE, asyncEstimate=True)
+        self.assertAsyncEstimation(task, FaceDetection)
+        task = self.defaultDetector.detect([VLIMAGE_ONE_FACE], asyncEstimate=True)
```
Comment (gluhar2006) [inline tests/test_facedetector.py]: for all tests with batches: it seems that it needs more than one object?

File: lunavl/sdk/async_task.py [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+        """
+        return postProcessingBatch(error, coreEstimations, resultClass=self.resultClass)
+
+    def postProcessingBatchWithAggregation(
+        self,
+        error: FSDKErrorResult,
+        coreEstimations: List[CoreTaskResult],
+        aggregatedAttribute: CoreTaskResult,
+        aggregate: bool,
+    ) -> tuple[List[TaskResult], Optional[TaskResult]]:
```
Comment (gluhar2006) [inline lunavl/sdk/async_task.py]: one more time

File: lunavl/sdk/async_task.py (hunk 14 above)
Comment (matemax) [reply lunavl/sdk/async_task.py:50]: double error validate?

File: pyproject.toml (hunk 11 above)
Comment (matemax) [reply pyproject.toml]: ++

File: pyproject.toml (hunk 12 above)
Comment (matemax) [reply pyproject.toml]: ++

File: lunavl/sdk/async_task.py (hunk 13 above)
Comment (matemax) [reply lunavl/sdk/async_task.py]: ++

File: lunavl/sdk/async_task.py (hunk 15 above)
Comment (matemax) [reply lunavl/sdk/async_task.py:57]: ++

File: lunavl/sdk/estimators/image_estimators/orientation_mode.py (hunk 22 above)
Comment (matemax) [reply lunavl/sdk/estimators/image_estimators/orientation_mode.py]: ++

File: lunavl/sdk/luna_faces.py (hunk 24 above)
Comment (matemax) [reply lunavl/sdk/luna_faces.py]: ++

File: tests/test_facedetector.py (hunk 26 above)
Comment (matemax) [reply tests/test_facedetector.py]: ++

File: lunavl/sdk/async_task.py (hunk 27 above)
Comment (matemax) [reply lunavl/sdk/async_task.py]: ++

File: tests/base.py (hunk 25 above)
Comment (matemax) [reply tests/base.py:166]: i don't understand you

File: lunavl/sdk/async_task.py [hunk 28]
```diff
@@ -0,0 +1,171 @@
...
+        aggregatedAttribute: aggregated core estimation
+        aggregate: need or not aggregate result
+        resultClass: result class
+
+    Returns:
+        list of lunavl structure based on core estimation + aggregated estimation
+    """
+    assertError(error)
+
+    estimations = [resultClass(coreEstimation) for coreEstimation in coreEstimations]
```
Comment (gluhar2006) [inline lunavl/sdk/async_task.py:50]:
```suggestion
    estimations = postProcessingBatch(error, coreEstimations, resultClass)
```

File: tests/base.py (hunk 25 above)
Comment (gluhar2006) [reply tests/base.py:166]: function named `assertAsyncBatchEstimationWithAggregation` - there is no mentions about batch size

**Pull request merged**